### ISSUE TITLE
wasm: segmentation fault remote wasm in ingressgateway

### ISSUE THREAD

**Issue by zufardhiyaulhaq** [istio/istio#29839]
**Bug description**
most of the tutorials using the local wasm file that mounted in the sidecar. Since not all my applications have Istio sidecar but exposed via Istio ingress gateway, my only options are to enable WASM filter in the ingress gateway.

after spikes, I found that annotation to mounted hostPath volume to a sidecar is not working in ingress gateway (?).  Since I already check `/var/local/lib/wasme-cache` on Kubernetes worker node and Ingress gateway directly (empty on ingress gateway pod)
```
sidecar.istio.io/userVolume: '[{"name":"cache-dir","hostPath":{"path":"/var/local/lib/wasme-cache"}}]'
sidecar.istio.io/userVolumeMount: '[{"mountPath":"/var/local/lib/wasme-cache","name":"cache-dir"}]'
```

I am checking via Envoy documentation and found out that we can add the wasm file remotely. https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/wasm/v3/wasm.proto#envoy-v3-api-msg-extensions-wasm-v3-vmconfig

my envoyfilter configuration
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: github-se
  namespace: istio-system
spec:
  hosts:
  - github.com
  location: MESH_EXTERNAL
  ports:
  - number: 443
    name: https
    protocol: TLS
  resolution: DNS
  exportTo:
    - "."

---
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: wasm-secure-header-public-gateway-1.7-ef
  namespace: istio-system
spec:
  workloadSelector:
    labels:
      istio: ingressgateway
  configPatches:
    - applyTo: HTTP_FILTER
      match:
        context: GATEWAY
        listener:
          filterChain:
            filter:
              name: envoy.http_connection_manager
              subFilter:
                name: envoy.router
        proxy:
          proxyVersion: ^1\.7.*
      patch:
        operation: INSERT_BEFORE
        value:
          name: envoy.filters.http.wasm
          typedConfig:
            '@type': type.googleapis.com/udpa.type.v1.TypedStruct
            typeUrl: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
            value:
              config:
                name: wasm-secure-header-public-gateway-1.7
                rootId: root_id
                vmConfig:
                  code:
                    remote:
                      http_uri:
                        uri: https://github.com/zufardhiyaulhaq/wasm-secure-header/raw/master/assets/v0.1.1/filter.wasm
                        cluster: outbound|443||github.com
                        timeout: 1s
                      sha256: d339aa2c407ff5db392cce4534b6232b9d732a764e4329caac587919ffd30ba3
                  runtime: envoy.wasm.runtime.v8
                  vmId: wasm-secure-header-public-gateway-1.7
```

I am writing the WASM filter in Go, please check https://github.com/zufardhiyaulhaq/wasm-secure-header

Istio ingress gateway logs
```
2021-01-02T19:02:39.612410Z     info    sds     resource:ROOTCA pushed root cert to proxy
2021-01-02T19:02:39.731425Z     critical        envoy backtrace Caught Segmentation fault, suspect faulting address 0x28
2021-01-02T19:02:39.731455Z     critical        envoy backtrace Backtrace (use tools/stack_decode.py to get line numbers):
2021-01-02T19:02:39.731459Z     critical        envoy backtrace Envoy version: b45c635f192a91ffbfdca5730d4198a3ec7827a3/1.15.1-dev/Clean/RELEASE/BoringSSL
2021-01-02T19:02:39.731847Z     critical        envoy backtrace #0: __restore_rt [0x7f8bcbc13980]
2021-01-02T19:02:39.815929Z     critical        envoy backtrace #1: Envoy::Extensions::Common::Wasm::createWasm() [0x557356d721dd]
2021-01-02T19:02:39.823403Z     critical        envoy backtrace #2: Envoy::Extensions::HttpFilters::Wasm::FilterConfig::FilterConfig() [0x55735684d6bb]
2021-01-02T19:02:39.830698Z     critical        envoy backtrace #3: Envoy::Extensions::HttpFilters::Wasm::WasmFilterConfig::createFilterFactoryFromProtoTyped() [0x55735684bb2e]
2021-01-02T19:02:39.837700Z     critical        envoy backtrace #4: Envoy::Extensions::HttpFilters::Common::FactoryBase<>::createFilterFactoryFromProto() [0x55735684be4f]
2021-01-02T19:02:39.921452Z     critical        envoy backtrace #5: Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerConfig::processFilter() [0x557357b4ba68]
2021-01-02T19:02:40.010990Z     critical        envoy backtrace #6: Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerConfig::HttpConnectionManagerConfig() [0x557357b4a4c4]
2021-01-02T19:02:40.025080Z     critical        envoy backtrace #7: Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProtoTyped() [0x557357b48422]
2021-01-02T19:02:40.116925Z     critical        envoy backtrace #8: Envoy::Extensions::NetworkFilters::Common::FactoryBase<>::createFilterFactoryFromProto() [0x557357b4e436]
2021-01-02T19:02:40.126943Z     critical        envoy backtrace #9: Envoy::Server::ProdListenerComponentFactory::createNetworkFilterFactoryList_() [0x557357b1266f]
2021-01-02T19:02:40.135954Z     critical        envoy backtrace #10: Envoy::Server::ProdListenerComponentFactory::createNetworkFilterFactoryList() [0x557357b1b304]
2021-01-02T19:02:40.219387Z     critical        envoy backtrace #11: Envoy::Server::ListenerFilterChainFactoryBuilder::buildFilterChainInternal() [0x557357b1aeff]
2021-01-02T19:02:40.233705Z     critical        envoy backtrace #12: Envoy::Server::ListenerFilterChainFactoryBuilder::buildFilterChain() [0x557357b1ad59]
2021-01-02T19:02:40.322041Z     critical        envoy backtrace #13: Envoy::Server::FilterChainManagerImpl::addFilterChain() [0x557357b2538c]
2021-01-02T19:02:40.334585Z     critical        envoy backtrace #14: Envoy::Server::ListenerImpl::buildFilterChains() [0x557357b0ccdb]
2021-01-02T19:02:40.416521Z     critical        envoy backtrace #15: Envoy::Server::ListenerImpl::ListenerImpl() [0x557357b0adf6]
2021-01-02T19:02:40.427646Z     critical        envoy backtrace #16: Envoy::Server::ListenerManagerImpl::addOrUpdateListenerInternal() [0x557357b16dd6]
2021-01-02T19:02:40.522561Z     critical        envoy backtrace #17: Envoy::Server::ListenerManagerImpl::addOrUpdateListener() [0x557357b164d2]
2021-01-02T19:02:40.629717Z     critical        envoy backtrace #18: Envoy::Server::LdsApiImpl::onConfigUpdate() [0x557357b44c75]
2021-01-02T19:02:40.713083Z     critical        envoy backtrace #19: Envoy::Server::LdsApiImpl::onConfigUpdate() [0x557357b45a6c]
2021-01-02T19:02:40.721366Z     critical        envoy backtrace #20: Envoy::Config::GrpcSubscriptionImpl::onConfigUpdate() [0x557357d385a9]
2021-01-02T19:02:40.729971Z     critical        envoy backtrace #21: Envoy::Config::GrpcMuxImpl::onDiscoveryResponse() [0x557357d3b398]
2021-01-02T19:02:40.818657Z     critical        envoy backtrace #22: Envoy::Grpc::AsyncStreamCallbacks<>::onReceiveMessageRaw() [0x557357bcff1f]
2021-01-02T19:02:40.826704Z     critical        envoy backtrace #23: Envoy::Grpc::AsyncStreamImpl::onData() [0x557357d783a7]
2021-01-02T19:02:40.834488Z     critical        envoy backtrace #24: Envoy::Http::AsyncStreamImpl::encodeData() [0x557357d7e1e8]
2021-01-02T19:02:40.916340Z     critical        envoy backtrace #25: Envoy::Router::UpstreamRequest::decodeData() [0x557357d9268e]
2021-01-02T19:02:40.928701Z     critical        envoy backtrace #26: Envoy::Http::ResponseDecoderWrapper::decodeData() [0x557357cfe69b]
2021-01-02T19:02:41.018534Z     critical        envoy backtrace #27: Envoy::Http::Http2::ConnectionImpl::onFrameReceived() [0x557357e1a0ff]
2021-01-02T19:02:41.112113Z     critical        envoy backtrace #28: nghttp2_session_on_data_received [0x557357ff5cfc]
2021-01-02T19:02:41.211021Z     critical        envoy backtrace #29: nghttp2_session_mem_recv [0x557357ff7e14]
2021-01-02T19:02:41.224793Z     critical        envoy backtrace #30: Envoy::Http::Http2::ConnectionImpl::innerDispatch() [0x557357e19699]
2021-01-02T19:02:41.312428Z     critical        envoy backtrace #31: std::__1::__function::__func<>::operator()() [0x557357e20f53]
2021-01-02T19:02:41.326186Z     critical        envoy backtrace #32: Envoy::Http::Utility::exceptionToStatus() [0x557357ea1da7]
2021-01-02T19:02:41.338196Z     critical        envoy backtrace #33: Envoy::Http::Http2::ConnectionImpl::dispatch() [0x557357e194b1]
2021-01-02T19:02:41.421070Z     critical        envoy backtrace #34: Envoy::Http::CodecClient::onData() [0x557357d6df10]
2021-01-02T19:02:41.431179Z     critical        envoy backtrace #35: Envoy::Http::CodecClient::CodecReadFilter::onData() [0x557357d6ed6d]
2021-01-02T19:02:41.522257Z     critical        envoy backtrace #36: Envoy::Network::FilterManagerImpl::onContinueReading() [0x557357b87f43]
2021-01-02T19:02:41.614797Z     critical        envoy backtrace #37: Envoy::Network::ConnectionImpl::onReadReady() [0x557357b8445e]
2021-01-02T19:02:41.627209Z     critical        envoy backtrace #38: Envoy::Network::ConnectionImpl::onFileEvent() [0x557357b8355d]
2021-01-02T19:02:41.712130Z     critical        envoy backtrace #39: Envoy::Event::FileEventImpl::assignEvents()::$_1::__invoke() [0x557357b7d7f6]
2021-01-02T19:02:41.719621Z     critical        envoy backtrace #40: event_process_active_single_queue [0x557357fe7aeb]
2021-01-02T19:02:41.727267Z     critical        envoy backtrace #41: event_base_loop [0x557357fe636e]
2021-01-02T19:02:41.734092Z     critical        envoy backtrace #42: Envoy::Server::InstanceImpl::run() [0x557357afd58f]
2021-01-02T19:02:41.816946Z     critical        envoy backtrace #43: Envoy::MainCommonBase::run() [0x557356564cb8]
2021-01-02T19:02:41.824187Z     critical        envoy backtrace #44: Envoy::MainCommon::main() [0x557356565495]
2021-01-02T19:02:41.831554Z     critical        envoy backtrace #45: main [0x5573565636fc]
2021-01-02T19:02:41.831686Z     critical        envoy backtrace #46: __libc_start_main [0x7f8bcb831bf7]
AsyncClient 0x557359df2400, stream_id_: 11012516632261428648
&stream_info_: 
  StreamInfoImpl 0x557359df25b0, protocol_: 1, response_code_: null, response_code_details_: via_upstream, health_check_request_: 0, route_name_:
2021-01-02T19:02:42.024415Z     info    sds     resource:ROOTCA connection is terminated: rpc error: code = Canceled desc = context canceled
2021-01-02T19:02:42.024465Z     error   sds     Remote side closed connection
2021-01-02T19:02:42.024541Z     error   Epoch 0 exited with error: signal: segmentation fault (core dumped)
2021-01-02T19:02:42.024553Z     info    No more active epochs, terminating
2021-01-02T19:02:42.024542Z     info    sds     resource:default connection is terminated: rpc error: code = Canceled desc = context canceled
```

I also try to use default WASM (just add hello: world) headers but found the same error.

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**
Istio ingress gateway pod not crashing, WASM function works

**Steps to reproduce the bug**
Applying the Envoyfilter make Istio ingress gateway pod crash

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)
$ istioctl version --remote
client version: 1.7.6
control plane version: 1.7.6
data plane version: 1.6.14 (5 proxies)

$ kubectl version --short
Client Version: v1.18.7
Server Version: v1.18.7

**How was Istio installed?**
Istioctl

**Environment where the bug was observed (cloud vendor, OS, etc)**
Kubernetes the hard way, OS Ubuntu 18.04

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

**Comment by zufardhiyaulhaq** [comment]
**Update**

I try to change using local files (using filterdeployment object) from wasme controller.
```
apiVersion: wasme.io/v1
kind: FilterDeployment
metadata:
  name: secure-header-helloworld-filterdeployment
  namespace: istio-testing
spec:
  deployment:
    istio:
      kind: Deployment
      labels:
        app: helloworld
  filter:
    image: webassemblyhub.io/zufardhiyaulhaq/secure-header:v0.1.1
```

the envoyfilter created:
```
spec:
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: SIDECAR_INBOUND
      listener:
        filterChain:
          filter:
            name: envoy.http_connection_manager
            subFilter:
              name: envoy.router
    patch:
      operation: INSERT_BEFORE
      value:
        name: envoy.filters.http.wasm
        typedConfig:
          '@type': type.googleapis.com/udpa.type.v1.TypedStruct
          typeUrl: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
          value:
            config:
              name: secure-header-helloworld-filterdeployment.istio-testing
              rootId: root_id
              vmConfig:
                code:
                  local:
                    filename: /var/local/lib/wasme-cache/d339aa2c407ff5db392cce4534b6232b9d732a764e4329caac587919ffd30ba3
                runtime: envoy.wasm.runtime.v8
                vmId: secure-header-helloworld-filterdeployment.istio-testing
  workloadSelector:
    labels:
      app: helloworld
```

its success running the wasm to add additional headers
```
< x-powered-by: Express
< content-type: application/json; charset=utf-8
< content-length: 1079
< etag: W/"437-qtdOg6owZOUkjLqK28GSUjk7Nlc"
< date: Sun, 03 Jan 2021 09:48:01 GMT
< x-envoy-upstream-service-time: 13
< x-frame-options: deny
< x-xss-protection: 1; mode=block
< x-content-type-options: nosniff
< referrer-policy: no-referrer
< x-download-options: noopen
< x-dns-prefetch-control: off
< strict-transport-security: max-age=31536000; includeSubDomains
< content-security-policy: frame-ancestors none;
< permissions-policy: camera 'none';microphone 'none';geolocation 'none';encrypted-media 'none';payment 'none';speaker 'none';usb 'none';
```

**Comment by zufardhiyaulhaq** [comment]
Update:

switching from local to remote make it crash
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: github-se
  namespace: istio-testing
spec:
  hosts:
  - github.com
  location: MESH_EXTERNAL
  ports:
  - number: 443
    name: https
    protocol: TLS
  resolution: DNS
  exportTo:
    - "."
---
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: wasm-secure-header-helloworld-ef
  namespace: istio-testing
spec:
  configPatches:
  - applyTo: HTTP_FILTER
    match:
      context: SIDECAR_INBOUND
      listener:
        filterChain:
          filter:
            name: envoy.http_connection_manager
            subFilter:
              name: envoy.router
    patch:
      operation: INSERT_BEFORE
      value:
        name: envoy.filters.http.wasm
        typedConfig:
          "@type": type.googleapis.com/udpa.type.v1.TypedStruct
          typeUrl: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
          value:
            config:
              name: wasm-secure-header-helloworld-ef
              rootId: root_id
              vmConfig:
                code:
                  remote:
                    http_uri:
                      uri: https://github.com/zufardhiyaulhaq/wasm-secure-header/raw/master/assets/v0.1.1/filter.wasm
                      cluster: outbound|443||github.com
                      timeout: 10s
                    sha256: d339aa2c407ff5db392cce4534b6232b9d732a764e4329caac587919ffd30ba3
                runtime: envoy.wasm.runtime.v8
                vmId: wasm-secure-header-helloworld-ef
  workloadSelector:
    labels:
      app: helloworld
```

```
2021-01-03T10:23:14.358626Z     critical        envoy backtrace Caught Segmentation fault, suspect faulting address 0x28
2021-01-03T10:23:14.358645Z     critical        envoy backtrace Backtrace (use tools/stack_decode.py to get line numbers):
2021-01-03T10:23:14.358671Z     critical        envoy backtrace Envoy version: b45c635f192a91ffbfdca5730d4198a3ec7827a3/1.15.1-dev/Clean/RELEASE/BoringSSL
2021-01-03T10:23:14.358855Z     critical        envoy backtrace #0: __restore_rt [0x7f7614c2c980]
2021-01-03T10:23:14.365490Z     critical        envoy backtrace #1: Envoy::Extensions::Common::Wasm::createWasm() [0x559057a3b1dd]
2021-01-03T10:23:14.373467Z     critical        envoy backtrace #2: Envoy::Extensions::HttpFilters::Wasm::FilterConfig::FilterConfig() [0x5590575166bb]
2021-01-03T10:23:14.380905Z     critical        envoy backtrace #3: Envoy::Extensions::HttpFilters::Wasm::WasmFilterConfig::createFilterFactoryFromProtoTyped() [0x559057514b2e]
2021-01-03T10:23:14.387855Z     critical        envoy backtrace #4: Envoy::Extensions::HttpFilters::Common::FactoryBase<>::createFilterFactoryFromProto() [0x559057514e4f]
2021-01-03T10:23:14.395936Z     critical        envoy backtrace #5: Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerConfig::processFilter() [0x559058814a68]
2021-01-03T10:23:14.403191Z     critical        envoy backtrace #6: Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerConfig::HttpConnectionManagerConfig() [0x5590588134c4]
2021-01-03T10:23:14.417680Z     critical        envoy backtrace #7: Envoy::Extensions::NetworkFilters::HttpConnectionManager::HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProtoTyped() [0x559058811422]
2021-01-03T10:23:14.425109Z     critical        envoy backtrace #8: Envoy::Extensions::NetworkFilters::Common::FactoryBase<>::createFilterFactoryFromProto() [0x559058817436]
2021-01-03T10:23:14.432931Z     critical        envoy backtrace #9: Envoy::Server::ProdListenerComponentFactory::createNetworkFilterFactoryList_() [0x5590587db66f]
2021-01-03T10:23:14.440907Z     critical        envoy backtrace #10: Envoy::Server::ProdListenerComponentFactory::createNetworkFilterFactoryList() [0x5590587e4304]
2021-01-03T10:23:14.449496Z     critical        envoy backtrace #11: Envoy::Server::ListenerFilterChainFactoryBuilder::buildFilterChainInternal() [0x5590587e3eff]
2021-01-03T10:23:14.458022Z     critical        envoy backtrace #12: Envoy::Server::ListenerFilterChainFactoryBuilder::buildFilterChain() [0x5590587e3d59]
2021-01-03T10:23:14.465938Z     critical        envoy backtrace #13: Envoy::Server::FilterChainManagerImpl::addFilterChain() [0x5590587ee38c]
2021-01-03T10:23:14.474666Z     critical        envoy backtrace #14: Envoy::Server::ListenerImpl::buildFilterChains() [0x5590587d5cdb]
2021-01-03T10:23:14.481958Z     critical        envoy backtrace #15: Envoy::Server::ListenerImpl::ListenerImpl() [0x5590587d3df6]
2021-01-03T10:23:14.489057Z     critical        envoy backtrace #16: Envoy::Server::ListenerManagerImpl::addOrUpdateListenerInternal() [0x5590587dfdd6]
2021-01-03T10:23:14.496799Z     critical        envoy backtrace #17: Envoy::Server::ListenerManagerImpl::addOrUpdateListener() [0x5590587df4d2]
2021-01-03T10:23:14.504065Z     critical        envoy backtrace #18: Envoy::Server::LdsApiImpl::onConfigUpdate() [0x55905880dc75]
2021-01-03T10:23:14.511744Z     critical        envoy backtrace #19: Envoy::Server::LdsApiImpl::onConfigUpdate() [0x55905880ea6c]
2021-01-03T10:23:14.518755Z     critical        envoy backtrace #20: Envoy::Config::GrpcSubscriptionImpl::onConfigUpdate() [0x559058a015a9]
2021-01-03T10:23:14.526263Z     critical        envoy backtrace #21: Envoy::Config::GrpcMuxImpl::onDiscoveryResponse() [0x559058a04398]
2021-01-03T10:23:14.534283Z     critical        envoy backtrace #22: Envoy::Grpc::AsyncStreamCallbacks<>::onReceiveMessageRaw() [0x559058898f1f]
2021-01-03T10:23:14.541590Z     critical        envoy backtrace #23: Envoy::Grpc::AsyncStreamImpl::onData() [0x559058a413a7]
2021-01-03T10:23:14.548811Z     critical        envoy backtrace #24: Envoy::Http::AsyncStreamImpl::encodeData() [0x559058a471e8]
2021-01-03T10:23:14.555412Z     critical        envoy backtrace #25: Envoy::Router::UpstreamRequest::decodeData() [0x559058a5b68e]
2021-01-03T10:23:14.561955Z     critical        envoy backtrace #26: Envoy::Http::ResponseDecoderWrapper::decodeData() [0x5590589c769b]
2021-01-03T10:23:14.568912Z     critical        envoy backtrace #27: Envoy::Http::Http2::ConnectionImpl::onFrameReceived() [0x559058ae30ff]
2021-01-03T10:23:14.575804Z     critical        envoy backtrace #28: nghttp2_session_on_data_received [0x559058cbecfc]
2021-01-03T10:23:14.583742Z     critical        envoy backtrace #29: nghttp2_session_mem_recv [0x559058cc0e14]
2021-01-03T10:23:14.590378Z     critical        envoy backtrace #30: Envoy::Http::Http2::ConnectionImpl::innerDispatch() [0x559058ae2699]
2021-01-03T10:23:14.597646Z     critical        envoy backtrace #31: std::__1::__function::__func<>::operator()() [0x559058ae9f53]
2021-01-03T10:23:14.604248Z     critical        envoy backtrace #32: Envoy::Http::Utility::exceptionToStatus() [0x559058b6ada7]
2021-01-03T10:23:14.610819Z     critical        envoy backtrace #33: Envoy::Http::Http2::ConnectionImpl::dispatch() [0x559058ae24b1]
2021-01-03T10:23:14.617241Z     critical        envoy backtrace #34: Envoy::Http::CodecClient::onData() [0x559058a36f10]
2021-01-03T10:23:14.624481Z     critical        envoy backtrace #35: Envoy::Http::CodecClient::CodecReadFilter::onData() [0x559058a37d6d]
2021-01-03T10:23:14.631857Z     critical        envoy backtrace #36: Envoy::Network::FilterManagerImpl::onContinueReading() [0x559058850f43]
2021-01-03T10:23:14.638723Z     critical        envoy backtrace #37: Envoy::Network::ConnectionImpl::onReadReady() [0x55905884d45e]
2021-01-03T10:23:14.645520Z     critical        envoy backtrace #38: Envoy::Network::ConnectionImpl::onFileEvent() [0x55905884c55d]
2021-01-03T10:23:14.653121Z     critical        envoy backtrace #39: Envoy::Event::FileEventImpl::assignEvents()::$_1::__invoke() [0x5590588467f6]
2021-01-03T10:23:14.660369Z     critical        envoy backtrace #40: event_process_active_single_queue [0x559058cb0aeb]
2021-01-03T10:23:14.667375Z     critical        envoy backtrace #41: event_base_loop [0x559058caf36e]
2021-01-03T10:23:14.673896Z     critical        envoy backtrace #42: Envoy::Server::InstanceImpl::run() [0x5590587c658f]
2021-01-03T10:23:14.680788Z     critical        envoy backtrace #43: Envoy::MainCommonBase::run() [0x55905722dcb8]
2021-01-03T10:23:14.687317Z     critical        envoy backtrace #44: Envoy::MainCommon::main() [0x55905722e495]
2021-01-03T10:23:14.693741Z     critical        envoy backtrace #45: main [0x55905722c6fc]
2021-01-03T10:23:14.693853Z     critical        envoy backtrace #46: __libc_start_main [0x7f761484abf7]
AsyncClient 0x55905aaa8400, stream_id_: 2949995910722546051
&stream_info_: 
  StreamInfoImpl 0x55905aaa85b0, protocol_: 1, response_code_: null, response_code_details_: via_upstream, health_check_request_: 0, route_name_: 
2021-01-03T10:23:14.782828Z     info    sds     resource:default connection is terminated: rpc error: code = Canceled desc = context canceled
```

**Comment by mandarjog** [comment]
Remote load is not recommended for production use.
The remote loader in 1.7 has known issues dealing with 302s. Specifically github.com/raw redirects (302) to raw.githubusercontent.com which is not handled correctly.

Please place the file in a gcs bucket, or use https://raw.githubusercontent.com/zufardhiyaulhaq/wasm-secure-header/master/assets/v0.1.1/filter.wasm 

For production use, you *should* use a local load. In Istio 1.9 timeframe there will be other ways of safely loading the module.

**Comment by zufardhiyaulhaq** [comment]
Reopen this issue,

@mandarjog currently testing in 1.7.6, cannot use
https://raw.githubusercontent.com/zufardhiyaulhaq/wasm-secure-header/master/assets/v0.1.1/filter.wasm.

I don't have $ to place the file in the GCS bucket, so I try to use Minio, still not working at all
https://minio.zufardhiyaulhaq.com/publics/istio/filter.wasm 
```
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: minio-zufardhiyaulhaq-com-se
  namespace: istio-system
spec:
  hosts:
  - minio.zufardhiyaulhaq.com
  location: MESH_EXTERNAL
  ports:
  - number: 443
    name: https
    protocol: TLS
  resolution: DNS
  exportTo:
    - "."
---
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: wasm-secure-header-public-gateway-1.7-ef
  namespace: istio-system
spec:
  workloadSelector:
    labels:
      istio: ingressgateway
  configPatches:
    - applyTo: HTTP_FILTER
      match:
        context: GATEWAY
        listener:
          filterChain:
            filter:
              name: envoy.http_connection_manager
              subFilter:
                name: envoy.router
        proxy:
          proxyVersion: ^1\.7.*
      patch:
        operation: INSERT_BEFORE
        value:
          name: envoy.filters.http.wasm
          typedConfig:
            "@type": type.googleapis.com/udpa.type.v1.TypedStruct
            typeUrl: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
            value:
              config:
                name: wasm-secure-header-public-gateway-1.7
                rootId: root_id
                vmConfig:
                  code:
                    remote:
                      http_uri:
                        uri: https://minio.zufardhiyaulhaq.com/publics/istio/filter.wasm 
                        cluster: outbound|443||minio.zufardhiyaulhaq.com
                        timeout: 10s
                      sha256: d339aa2c407ff5db392cce4534b6232b9d732a764e4329caac587919ffd30ba3
                  runtime: envoy.wasm.runtime.v8
                  vmId: wasm-secure-header-public-gateway-1.7
```

**Comment by bianpengyuan** [comment]
An experimental Wasm module distribution is available at Istio 1.9 now, please take a look: https://istio.io/latest/docs/ops/configuration/extensibility/wasm-module-distribution/ thanks!
